### PR TITLE
feat(auth): multi-provider OAuth account linking (#517)

### LINKED ISSUE #517: feat(auth): multi-provider account linking (Connect Google + GitHub)
## Context

Spawned from #481 gate1 review and post-ship operator discussion (2026-04-30). #481 explicitly punted cross-provider linking as **Out of Scope** because it requires a schema change and a new OAuth-flow mode. After #481 ships, a Google-registered user attempting GitHub OAuth login with the same email is rejected with `ConflictError(409)` → `/login?error=email_in_use`. This issue tracks adding **opt-in** linking so one user account can be bound to multiple OAuth providers.

## Problem

Today the OAuth identity model is **1 user ↔ 1 provider**:
- `users.user_id` = OAuth `sub` (immutable, unique).
- `users.auth_provider` = `'google' | 'github' | 'password'`, set at creation only.

Real users frequently:
- Use Google at home, GitHub at work, expect one account
- Want to migrate from one provider to another without losing memory continuity
- Need a fallback auth path if one IdP is down

The 409 path from #481 is the correct **default** (security-first), but advanced users should be able to deliberately link an additional provider to their existing account.

## Scope

**In scope:**
- Schema: new table `user_oauth_providers (user_id FK, provider, oauth_sub, linked_at, last_used_at)` with composite UNIQUE on `(provider, oauth_sub)` — preserves the current invariant that an OAuth identity belongs to at most one user.
- Migrate existing `users.user_id` + `users.auth_provider` rows into `user_oauth_providers` as part of the migration (1 row per existing user).
- New endpoint: `POST /api/v1/me/account/link-provider` — initiates an OAuth flow in "link mode" (caller must already have a session; the new provider's `sub` is bound to the session user).
- New endpoint: `POST /api/v1/me/account/unlink-provider` — reject if it would leave the account with zero auth methods.
- Profile UI: "Connected accounts" section showing each linked provider + Connect / Disconnect buttons (sibling of #514's `auth_provider` display, supersedes once shipped).
- ConflictError on cross-provider OAuth login → redirect to `/login?error=email_in_use&link_hint=true` with a UI prompt: "An account exists with this email. Sign in with [Google] to link your [GitHub] account."
- Audit: `oauth_provider_linked` and `oauth_provider_unlinked` actions in `audit_logs`.
- i18n strings (en + ja).

**Out of scope (separate issues):**
- Multi-email per user (one user account, multiple email addresses)
- Provider-specific feature flags (e.g., GitHub-only enterprise SSO)
- Auto-linking by email match (security risk — must be explicit)

## Acceptance

- [ ] Migration creates `user_oauth_providers` table and backfills all existing users
- [ ] `RoleManager.ensure_user` looks up by `(provider, oauth_sub)` instead of `user_id` directly
- [ ] `POST /me/account/link-provider` happy path: user with Google session links GitHub account, both providers can sign them in afterward
- [ ] `POST /me/account/unlink-provider` happy path: user with 2 providers unlinks one
- [ ] `POST /me/account/unlink-provider` blocked: user with only 1 provider (no fallback) → 409
- [ ] ConflictError flow: cross-provider OAuth attempt redirects with link_hint + UI prompts user to link
- [ ] Audit log entries on link/unlink with HMAC-keyed hashes (reuse `AUDIT_HMAC_KEY` from #481)
- [ ] Profile page "Connected accounts" section renders correctly
- [ ] Cannot link a provider already bound to a different user (composite UNIQUE constraint enforced)
- [ ] E2E (Playwright): full link → unlink → re-link flow

## Design notes

- Reuses `AUDIT_HMAC_KEY` from #481 — no new env var
- `users.user_id` and `users.auth_provider` columns deprecated but retained for one release for read compatibility (`users.auth_provider` becomes "primary" provider for backward-compat); remove in a follow-up after frontend migration
- The link-mode OAuth flow uses a session cookie + state token to bind the new provider's `sub` to the existing user — distinct from regular login flow
- Security: the link endpoint requires fresh re-auth (within last 5 min) — protects against session-hijack-then-link attacks

## Related

- #481 (parent — established the 409 reject path for cross-provider attempts)
- #514 (sibling — show single auth_provider; superseded by this issue's "Connected accounts" UI)
- #515 (sibling — manual re-sync; orthogonal, both can ship independently)

## Design clarifications from gate1 CSO review (2026-06-07)

Security edge cases to pin down before/during implementation (gate1 verdict: yellow):

- [ ] **Re-auth contract for OAuth-only users**: "fresh re-auth within 5 min" is defined for password-auth users (password re-prompt, cf. `me_account.py:46-89` erasure flow) but undefined for OAuth-only users (no password). Define concretely — e.g. the OAuth round-trip itself + `oauth2_state_user:{state}` binding serves as the freshness proof. Must not assume a password exists.
- [ ] **`ensure_user` lookup-switch sequencing**: switching the lookup key from `users.user_id` to `(provider, oauth_sub)` touches every login path (google/github/password callbacks + `refresh-oauth` + device flow #640). Sequence within one PR: backfill table → read from new table with `users.user_id` fallback → defer column removal. AC must spell out the order so reads never split.
- [ ] **link callback 3-arm logic (explicit, no 500)**: incoming `sub` is ① unbound → link; ② already bound to *me* → idempotent no-op; ③ bound to *another user* → clean **409** (never IntegrityError→500, never account merge). The composite UNIQUE guards the DB; the app layer must branch explicitly.
- [ ] **Migration backfill for legacy-null `auth_provider`**: pre-#361 users have `auth_provider = NULL` (cf. `me_oauth.py:152`). "1 row per user" backfill cannot create a row without a provider — decide handling (skip / sentinel / infer) before the migration.
- [ ] **CWE-601 on link_hint redirect**: route `/login?error=email_in_use&link_hint=true` through auth.py's `_safe_redirect_url` validator (Pattern B/C), same as the other callbacks.
- [ ] **Audit failed link attempts**: arm ③ (link rejected because `sub` belongs to another user) is a security event — audit the *attempt*, not only success/unlink.
- [ ] **Unlinking the "primary" provider**: during the one-release compat window, define behavior when unlinking the provider equal to `users.auth_provider` (avoid orphaning the legacy column).

**Implementation note**: link-mode OAuth flow should clone the existing Pattern A in `me_oauth.py` (`INTENT_KEY`/`USER_KEY` state binding, `STATE_TTL=300`) — do not invent a new state mechanism.

### PR DESCRIPTION
**Pull request by JFK**
Closes #517

## Summary
- One user account can opt-in link multiple OAuth providers (Google + GitHub); either IdP then signs the same account in. OAuth↔OAuth only — no email/password signup work.
- New `user_oauth_providers (provider, oauth_sub) → users.user_id` satellite table with composite `UNIQUE(provider, oauth_sub)` + `UNIQUE(user_id, provider)`; backfill migration (skips legacy-null / password users).
- `ensure_user` switched to a dual-read (resolve by `(provider, oauth_sub)` → legacy `user_id` fallback → self-heal insert), so a linked secondary provider resolves the owning account while existing logins never break.
- `POST /me/account/{link,unlink}-provider` + `GET /me/account/providers`, an OAuth callback link-mode branch (3-arm), and a "Connected accounts" profile section (en/ja).

## Implementation notes
- **Migration strategy (expand-contract)**: one PR creates the table + backfill, adds the dual-read with legacy fallback, and retains `users.user_id` / `users.auth_provider` for one release (removal tracked as a follow-up). Existing-user reads never split. Revision `e36_888 → e37_517`, single head, reversible.
- **link-mode = fresh re-auth**: `link-provider` clones the existing `me_oauth` Pattern A state binding (`oauth2_state_intent="link"`, `oauth2_state_user={session user_id}`, `STATE_TTL=300`). The OAuth round-trip itself is the re-auth; no password prompt (OAuth-only users have none).
- **3-arm link callback** (`_maybe_link_redirect`, shared with both google/github callbacks): unbound → link; already-mine → idempotent no-op; bound to another user → clean **409** (never merges accounts, `IntegrityError` race caught → 409, not 500). Runs before the signup gate / `ensure_user`, creates/rotates **no** session.
- **Security**: all redirects (incl. `?error=email_in_use&link_hint=true`) pass through `_safe_redirect_url` (CWE-601). Audit rows for link / unlink / **failed** link attempts, HMAC-hashed via `AUDIT_HMAC_KEY` (no plaintext). Unlink rejected (409) if it would leave zero auth methods; primary provider repointed on unlink.

## Pre-PR review summary
- gate2 mode: advisor-only
- audit: skipped
- binary_gate: (none)
- cso: green
- qa-lead: green
- cto: green
- gate1: yellow via /ask (CSO lens) — 7 security edge cases added to the issue AC, each mapped to a test
- review provider: code-review

## Tests
- backend: 142 passed (auth + linking + model + callback), migration integration 6 passed — all re-run-safe (uuid-suffixed).
- frontend: ConnectedAccounts 7, profile 23 (isolation); `tsc --noEmit` clean.
- Every gate1 security edge case (re-auth contract, lookup sequencing, 3-arm, legacy-null backfill, CWE-601, failed-attempt audit, primary repoint) is covered by a named test.

## Deferred follow-ups
1. OAuth test-IdP E2E harness for the full link→unlink→re-link Playwright flow (repo has no IdP stub; authed E2E not yet in CI — #785/#786).
2. Remove deprecated `users.auth_provider` read-dependence one release after self-heal saturates.

🤖 Generated via /gh-issue-driven:ship

### CODE REVIEW

File: backend/src/api/routes/auth.py [hunk 1]
```diff
@@ -303,6 +311,111 @@ async def _maybe_refresh_redirect(
...
+        async for db in get_db():
+            await AccountLinkingService(db).link(
+                user_id=user_id,
+                provider=provider,
+                oauth_sub=idp_sub,
+                email=idp_email,
+                ip_address=ip_address,
+                user_agent=user_agent,
+            )
+            break
```
[the remark this answers is not in the document]
Comment (JFK) [reply backend/src/api/routes/auth.py:408]: ✅ Fixed in 97cb84ff. `_maybe_link_redirect` now loads the initiating session User by `user_id` (from `oauth2_state_user`) and passes that user's email to `AccountLinkingService.link()` for the audit actor — no longer the IdP-returned email. If the session user row can't be loaded (deleted mid-flow), it redirects to `?error=link_failed` instead of linking. Added a regression test asserting the audit `user_email` is the actor's, not the IdP's.
